shared/ssl.sh

```diff
@@ -0,0 +1,117 @@
+[ -z "${PUPPETSERVER_HOSTNAME}" ] && error "PUPPETSERVER_HOSTNAME must be non-empty value"
+SSLDIR="${SSLDIR:-/etc/puppetlabs/puppet/ssl}"
+[ -z "${SSLDIR}" ] && error "SSLDIR must be non-empty value"
+WAITFORCERT=${WAITFORCERT:-120}
+# TODO check that WAITFORCERT is a number before continuing?
+
+### Create directories and files
+PUBKEYDIR="${SSLDIR}/public_keys"
+PRIVKEYDIR="${SSLDIR}/private_keys"
+CSRDIR="${SSLDIR}/certificate_requests"
+CERTDIR="${SSLDIR}/certs"
+mkdir -p "${SSLDIR}" "${PUBKEYDIR}" "${PRIVKEYDIR}" "${CSRDIR}" "${CERTDIR}"
+PUBKEYFILE="${PUBKEYDIR}/${CERTNAME}.pem"
+PRIVKEYFILE="${PRIVKEYDIR}/${CERTNAME}.pem"
+CSRFILE="${CSRDIR}/${CERTNAME}.pem"
+CERTFILE="${CERTDIR}/${CERTNAME}.pem"
+CACERTFILE="${CERTDIR}/ca.pem"
+
+CA="https://${PUPPETSERVER_HOSTNAME}:8140/puppet-ca/v1"
+CERTHEADER="-----BEGIN CERTIFICATE-----"
+
+### Print configuration for troubleshooting
+msg "Getting signed certificate for '${CERTNAME}'"
+msg "Using Puppet CA at '${CA}'"
+msg "Saving SSL files in '${SSLDIR}'"
+msg "Waiting up to ${WAITFORCERT} seconds for certificate to be signed"
+
+### Pre-checks before we begin:
+# 1) We can reach the CA host
+# 2) CA does not already have a signed certificate for this host
+output="$(curl --insecure --silent "${CA}/certificate/${CERTNAME}" | head -1)"
+if [ $? -ne 0 ]; then
+    error "cannot reach CA host '${PUPPETSERVER_HOSTNAME}'"
+elif [ "${output}" = "${CERTHEADER}" ]; then
+    error "CA already has signed certificate for '${CERTNAME}'"
+fi
+
+### Generate keys and CSR for this host
+[ -s "${PRIVKEYFILE}" ] && error "private key '${PRIVKEYFILE}' already exists"
+[ -s "${PUBKEYFILE}" ] && error "public key '${PUBKEYFILE}' already exists"
+[ -s "${CSRFILE}" ] && error "certificate request '${CSRFILE}' already exists"
+openssl genrsa -out "${PRIVKEYFILE}" 4096
+openssl rsa -in "${PRIVKEYFILE}" -pubout -out "${PUBKEYFILE}"
+openssl req -new -key "${PRIVKEYFILE}" -out "${CSRFILE}" -subj "/CN=${CERTNAME}"
+
+### Submit CSR
+curl --insecure -X PUT -H "Content-Type: text/plain" \
+    --data-binary "@${CSRFILE}" "${CA}/certificate_request/${CERTNAME}"
+
+### Retrieve signed certificate; wait and try again with a timeout
+sleeptime=10
+timewaited=0
+cert=$(curl --insecure "${CA}/certificate/${CERTNAME}")
+while [ "$(echo "${cert}" | head -1)" != "${CERTHEADER}" ]; do
+    [ ${timewaited} -ge ${WAITFORCERT} ] && \
+        error "timed-out waiting for certificate to be signed"
+    msg "Waiting for certificate to be signed..."
+    sleep ${sleeptime}
+    timewaited=$((${timewaited}+${sleeptime}))
+    cert="$(curl --insecure "${CA}/certificate/${CERTNAME}")"
+done
+echo "${cert}" > "${CERTFILE}"
+
+### Also get the CA certificate
+curl --insecure --output "${CACERTFILE}" "${CA}/certificate/ca"
+
+### Verify we got a signed certificate
+if [ -f "${CERTFILE}" ] && [ "$(head -1 "${CERTFILE}")" = "${CERTHEADER}" ]; then
+    msg "Successfully signed certificate '${CERTFILE}'"
```

Comment:
Maybe also check that openssl is happy with the file ? I am thinking of something like `openssl x509 -in $CERTFILE -subject -noout`

Reply:
Good call. Adding that.